Mathlib/LinearAlgebra/Matrix/Ideal.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Wojciech Nawrocki
+-/
+import Mathlib.Data.Matrix.Basis
+import Mathlib.RingTheory.TwoSidedIdeal.Operations
+
+/-!
+# Ideals in a matrix ring
+
+This file defines left (resp. two-sided) ideals in a matrix semiring (resp. ring)
+over left (resp. two-sided) ideals in the base semiring (resp. ring).
+
+## Main results
+
+* `TwoSidedIdeal.equivMatricesOver` and `TwoSidedIdeal.orderIsoMatricesOver`
+  establish an order isomorphism between two-sided ideals in `R` and those in `Mₙ(R)`.
+-/
+
+/-! ### Left ideals in a matrix ring -/
+
+namespace Ideal
+open Matrix
+
+variable {R : Type*} [Semiring R]
+         (n : Type*) [Fintype n] [DecidableEq n]
+
+/-- The left ideal of matrices with entries in `I ≤ R`. -/
+def matricesOver (I : Ideal R) : Ideal (Matrix n n R) where
+  carrier := { M | ∀ i j, M i j ∈ I }
+  add_mem' ha hb i j := I.add_mem (ha i j) (hb i j)
+  zero_mem' _ _ := I.zero_mem
+  smul_mem' M N hN := by
+    intro i j
+    rw [smul_eq_mul, mul_apply]
+    apply sum_mem
+    intro k _
+    apply I.mul_mem_left _ (hN k j)
+
+@[simp]
+theorem mem_matricesOver (I : Ideal R) (M : Matrix n n R) :
+    M ∈ I.matricesOver n ↔ ∀ i j, M i j ∈ I := by rfl
+
+def matricesOver_mono {I J : Ideal (Matrix n n R)} :
+    I ≤ J → I.matricesOver n ≤ J.matricesOver n :=
+  fun IJ _ MI i j => IJ (MI i j)
```

Comment:
Can you state this with `Monotone`? Presumably it's `StrictMono` if `Nonempty n`?

Reply:
Done (yeah it's `StrictMono`). Another way to do this would be to bundle `matricesOver` as an order hom, like [TwoSidedIdeal.asIdeal](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/TwoSidedIdeal/Operations.html#TwoSidedIdeal.asIdeal). I don't like this as much because dot notation no longer works, and there is a [CanLift instance](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Order/Hom/Basic.html#OrderHom.canLift) for maps that are known to be Monotone. But maybe there is something we lose by unbundling? Also if one is to bundle, I feel like it should be bundled as the strongest possible structure. It should at least be a `BotTopHom` (except that doesn't exist).